src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java

```diff
@@ -30,436 +29,446 @@
+        }
+        SwerveModuleState[] desaturatedStates = new SwerveModuleState[states.length];
+        for (int i = 0; i < states.length; i++) {
+            desaturatedStates[i] = new SwerveModuleState(states[i].speedMetersPerSecond / desaturationFactor, states[i].angle);
+        }
+        return desaturatedStates;
+    }
+
+    public ChassisSpeeds getChassisSpeeds() {
+        return kinematics.toChassisSpeeds(
+                getSwerveModuleStates()
+        );
+    }
+
+    public SwerveModulePosition[] getSwerveModulePositions() {
+        return new SwerveModulePosition[]{
+                frontLeft.getCurrentPosition(),
+                frontRight.getCurrentPosition(),
+                backLeft.getCurrentPosition(),
+                backRight.getCurrentPosition()
+        };
+    }
+
+    public SwerveModuleState[] getSwerveModuleStates(){
+        SwerveModuleState[] moduleStates = new SwerveModuleState[Module.values().length];
+        for (Module module : Module.values()) {
+            moduleStates[module.ordinal()] = getModuleState(module);
+        }
+        return moduleStates;
+    }
+
+    public SwerveDriveKinematics getKinematics() {
+        return this.kinematics;
+    }
+
+    public IBaseGyro getGyro() {
+        return gyro;
+    }
+
+    /**
+     * moving a single module by module state
+     */
+    private void setModuleStateForModule(Module module, SwerveModuleState state) {
+        getModule(module).setModuleState(state);
+    }
+
+    /**
+     * for calibration purposes
+     */
+    public void rotateModuleByPower(Module module, double power) {
+        getModule(module).setRotationalPower(power);
+    }
+
+    public void updatePoseEstimationPhotonVision() {
+        poseEstimator.update(getGyroAngle(), getSwerveModulePositions());
+        Photonvision.getInstance().getUpdatedPoseEstimator().ifPresent((EstimatedRobotPose pose) -> poseEstimator.addVisionMeasurement(pose.estimatedPose.toPose2d(), pose.timestampSeconds));
+    }
+
+    public void updatePoseEstimationLimeLight() {
+        boolean poseDifference = odometry.getPoseMeters().getTranslation().getDistance(getRobotPose().getTranslation()) < OdometryConstants.MAX_DISTANCE_TO_FILTER_OUT;
+        boolean shouldUpdateByOdometry = poseDifference && !isRobotNotOnGround();
+        if (shouldUpdateByOdometry) {
+            poseEstimator.update(getGyroAngle(), getSwerveModulePositions());
+        }
+        if (doVision) {
+            for (Optional<Pair<Pose2d, Double>> target : MultiLimelight.getInstance().getAllEstimates()) {
+                target.ifPresent(this::addVisionMeasurement);
+            }
+        }
+    }
+
+    private boolean isModuleAtFreeCurrent(SwerveModule module) {
+        return module.getLinearCurrent() > MK4iSwerveConstants.LINEAR_MOTOR_FREE_CURRENT - CURRENT_TOLERANCE && module.getLinearCurrent() < CURRENT_TOLERANCE + MK4iSwerveConstants.LINEAR_MOTOR_FREE_CURRENT;
+    }
+
+    public boolean isRobotNotOnGround() {
```

Comment:
I think it would be better to have a method for `isRobotOnGround` and use `!` on the result

Or at the very least change to `isRobotIsTheAir` or something like that